src/config/unit_test/mod.rs

```diff
@@ -0,0 +1,499 @@
+            ]);
+        }
+        let outputs = build_outputs(outputs)?;
+
+        let mut template_sinks = IndexMap::new();
+        let mut test_result_rxs = Vec::new();
+        // Add sinks with checks
+        for (id, checks) in outputs {
+            let (tx, rx) = oneshot::channel();
+            let sink_config = UnitTestSinkConfig {
+                test_name: test_name.to_string(),
+                transform_id: id.to_string(),
+                result_tx: Arc::new(Mutex::new(Some(tx))),
+                check: UnitTestSinkCheck::Checks(checks),
+            };
+
+            test_result_rxs.push(rx);
+            template_sinks.insert(id.clone(), sink_config);
+        }
+
+        // Add sinks with no outputs check
+        for id in no_outputs_from {
+            let (tx, rx) = oneshot::channel();
+            let sink_config = UnitTestSinkConfig {
+                test_name: test_name.to_string(),
+                transform_id: id.to_string(),
+                result_tx: Arc::new(Mutex::new(Some(tx))),
+                check: UnitTestSinkCheck::NoOutputs,
+            };
+
+            test_result_rxs.push(rx);
+            template_sinks.insert(id.clone(), sink_config);
+        }
+
+        let sinks = template_sinks
+            .into_iter()
+            .map(|(transform_id, sink_config)| {
+                let sink_id = self
+                    .sink_ids
+                    .get(&transform_id)
+                    .expect("Sink does not exist")
+                    .as_ref();
+                (
+                    ComponentKey::from(sink_id),
+                    SinkOuter::new(vec![transform_id.to_string()], Box::new(sink_config)),
+                )
+            })
+            .collect::<IndexMap<_, _>>();
+
+        Ok((test_result_rxs, sinks))
+    }
+}
+
+// Find all components that participate in the test
+fn get_relevant_test_components(
+    sources: &[&ComponentKey],
+    graph: &Graph,
+) -> Result<HashSet<String>, Vec<String>> {
+    let _ = graph.check_for_cycles().map_err(|error| vec![error])?;
+    Ok(sources
+        .iter()
+        .flat_map(|source| {
+            graph
+                .paths_to_sink_from(source)
+                .into_iter()
+                .map(|key| key.to_string())
+        })
+        .collect())
+}
+
+async fn build_unit_test(
+    metadata: &UnitTestBuildMetadata,
+    test: TestDefinition,
+    mut config_builder: ConfigBuilder,
+) -> Result<UnitTest, Vec<String>> {
+    let sources = metadata.hydrate_into_sources(&test.inputs)?;
+    let (test_result_rxs, sinks) =
+        metadata.hydrate_into_sinks(&test.name, &test.outputs, &test.no_outputs_from)?;
+
+    config_builder.sources = sources;
+    config_builder.sinks = sinks;
+
+    // To properly identify all components relevant to the test, expand relevant
+    // transforms
+    let mut expanded_config = config_builder.clone();
+    let _ = expand_macros(&mut expanded_config);
+
+    let graph = Graph::new_unchecked(
```

Comment:
Why do we need to use `new_unchecked` here?

Reply:
Good question -- it mainly has to do with the ordering in which we sanitize the config (i.e. remove unnecessary components) and transform inputs (i.e. remove invalid inputs). 

The graph is constructed before irrelevant transforms are removed and before transform inputs are sanitized, so extraneous transform inputs (e.g. user-configured sources) that no longer actually exist will lead to errors. For example, prior to building tests, in `UnitTestBuildMetadata::initialize()`, we add a test source as an input to all transforms. When actually building each test, we create only the sources we need based on test inputs. Thus, when creating the graph, we may include unnecessary transforms with non-existent test source inputs which leads to errors.